### PR TITLE
[DYNAMIC] Add Dynamic reshape to a dynamic namespace and add DynamicToStatic Pass

### PR DESCRIPTION
**Pull request by mbrookhart**
Initial POC for ideas discussed in https://discuss.tvm.ai/t/dynamic-ops-in-relay/6909

@tqchen @jroesch @kevinthesun @lixiaoquan @icemelon9

### CODE REVIEW

**Comment by mbrookhart** [comment]
Hitting an issue with dynamic shapes on GPU with the VM.

File: python/tvm/relay/op/dynamic/transform.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+        .. code-block:: python
+
+            data.shape = (2,3,4), newshape = (-3,4), result.shape = (6,4)
+            data.shape = (2,3,4,5), newshape = (-3,-3), result.shape = (6,20)
+            data.shape = (2,3,4), newshape = (0,-3), result.shape = (2,12)
+            data.shape = (2,3,4), newshape = (-3,-2), result.shape = (6,4)
+
+    Special values -2 and -4 from the standard reshape op would introduce dynamic rank
+    in this op. Thus, they are not permitted.
```
Comment (lixiaoquan) [inline python/tvm/relay/op/dynamic/transform.py:59]: Current code allows -2 and -4 as constant value. Should frontend check it like
```
if newshape_inferable() and newshape == (-2) or newshape == (-4):
    return _op.reshape()
else:
    return _op.dynamic.reshape()
```

Comment (mbrookhart) [reply python/tvm/relay/op/dynamic/transform.py:59]: The Input is a dynamic Expression, I can't check it at graph construction time.

Comment (mbrookhart) [reply python/tvm/relay/op/dynamic/transform.py:59]: It's a runtime Expr, I can't check it at compile time.

**Comment by icemelon9** [comment]
Shall we just call the namespace `dyn`? I'm okay with `dynamic`, but just feel it is a bit long.

**Comment by tqchen** [comment]
I agree with @icemelon9 perhaps `dyn` is a better name here.

**Comment by mbrookhart** [comment]
:+1: Happy to shorten it.

File: python/tvm/relay/op/dyn/transform.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+# pylint: disable=import-outside-toplevel
+"""Dynamic Transform operators."""
+
+from . import _make
+
+
+def reshape(data, newshape):
```
Comment (icemelon9) [inline python/tvm/relay/op/dyn/transform.py:24]: I suggest in the frontend we only keep one `reshape` and inside the frontend you can decide whether to use `dyn.reshape` or `reshape`. So this means that we remove this file and instead update the `reshape` in the op/transform.py

**Comment by icemelon9** [comment]
In fact, I have another thought on this. Should we make the dynamic ops as the default ones whereas having static ops as special cases for better optimization?

**Comment by jroesch** [comment]
@icemelon9 That was my initial suggestion, I think we reached consensus between TQ, Matt and I that we would introduce dynamic and slowly migrate stuff to use them, then introduce static migrate everything to that, and then finally remove the dynamic prefix inorder to ensure a smooth migration without confusion.

File: python/tvm/relay/op/dyn/transform.py (hunk 2 above)
Comment (mbrookhart) [reply python/tvm/relay/op/dyn/transform.py:24]: This is an interesting idea. It makes buy into the dynamic op namespace implicit, which reduces changes needed in the frontends later, but it also forces people to use the namespace earlier. @jroesch What do you think, is implicit or explicit better at this stage?

File: python/tvm/relay/op/dyn/transform.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    ``-3`` use the product of two consecutive dimensions of the input shape
+    as the output dimension.
+
+        .. code-block:: python
+
+            data.shape = (2,3,4), newshape = (-3,4), result.shape = (6,4)
+            data.shape = (2,3,4,5), newshape = (-3,-3), result.shape = (6,20)
+            data.shape = (2,3,4), newshape = (0,-3), result.shape = (2,12)
+            data.shape = (2,3,4), newshape = (-3,-2), result.shape = (6,4)
```
Comment (electriclilies) [inline python/tvm/relay/op/dyn/transform.py:56]: This example uses -2 as an argument (in the last example), but it was noted earlier that -2 is not allowed as an argument in dynamic reshape.

File: src/relay/op/dyn/tensor/transform.cc [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    ``-3`` use the product of two consecutive dimensions of the input shape
+    as the output dimension.
+
+        .. code-block:: python
+
+            data.shape = (2,3,4), newshape = (-3,4), result.shape = (6,4)
+            data.shape = (2,3,4,5), newshape = (-3,-3), result.shape = (6,20)
+            data.shape = (2,3,4), newshape = (0,-3), result.shape = (2,12)
+            data.shape = (2,3,4), newshape = (-3,-2), result.shape = (6,4)
```
Comment (electriclilies) [inline src/relay/op/dyn/tensor/transform.cc:119]: -2 is used as an argument in this example but -2 isn't allowed as an argument for dynamic reshape

File: python/tvm/relay/op/dyn/transform.py (hunk 3 above)
Comment (mbrookhart) [reply python/tvm/relay/op/dyn/transform.py]: Thanks!

**Comment by tqchen** [comment]
cc @kevinthesun @lixiaoquan given that we agreed on the new convention, it would be great if we can also migrate some of the existing dynamic(symbolic) effort towards this direction

**Comment by tqchen** [comment]
cc @icemelon9 please followup and manage this PR, @jroesch @electriclilies please https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly

File: python/tvm/relay/op/dyn/transform.py (hunk 2 above)
Comment (icemelon9) [reply python/tvm/relay/op/dyn/transform.py:24]: @jroesch Could you comment on this?

**Comment by icemelon9** [comment]
@mbrookhart Since we separate the static and dynamic reshape in this pr, we should remove the `Optional` in the `ReshapeAttrs` correspondingly.

**Comment by mbrookhart** [comment]
@icemelon9 @lixiaoquan Per Haichen's request, I removed the dynamic option from the standard reshape and changed the places that used it to use the new purely dynamic op. This touches many more files, could you take another look?

File: python/tvm/relay/op/dyn/transform.py (hunk 2 above)
Comment (jroesch) [reply python/tvm/relay/op/dyn/transform.py:24]: I think its okay to have a single interface in the Python code, at least I talked with Haichen about it offline. It seems less confusing for end users who want to build models, but will require that people understand the difference between the two ops in the backend (which is true either way).

Comment (mbrookhart) [reply python/tvm/relay/op/dyn/transform.py:24]: Okay, I'll rework the python API to merge them.

Comment (mbrookhart) [reply python/tvm/relay/op/dyn/transform.py]: I merged the two interfaces, let me know what you think.

File: src/relay/analysis/util.cc [hunk 5]
```diff
@@ -448,13 +448,8 @@ bool IsDataDependant(const CallNode* call) {
...
 
-  if (op->name == "reshape") {
-    if (const auto* attrs = call->attrs.as<ReshapeAttrs>()) {
-      if (attrs->newshape) {
-        // If newshape attribute exists, it isn't data dependant.
-        return false;
-      }
-    }
+  if (op->name == "dyn.reshape") {
+    return true;
```
Comment (icemelon9) [inline src/relay/analysis/util.cc:452]: We can remove line 451-452 as the shape function of `dyn.reshape` is registered to be data dependent.

File: python/tvm/relay/op/_tensor_grad.py [hunk 6]
```diff
@@ -511,7 +511,7 @@ def batch_matmul_grad(orig, grad):
 @register_gradient("reshape")
 def reshape_grad(orig, grad):
     """Gradient of reshape"""
-    return [reshape_like(grad, orig.args[0]), orig.args[1]]
+    return [reshape_like(grad, orig.args[0])]
```
Comment (icemelon9) [inline python/tvm/relay/op/_tensor_grad.py:514]: Do we need to register a gradient op to `dyn.reshape`?

File: src/relay/op/dyn/tensor/transform.cc [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+#include <tvm/relay/op.h>
+#include <tvm/relay/op_attr_types.h>
+#include <tvm/runtime/registry.h>
+
+namespace tvm {
+namespace relay {
+namespace dyn {
+
+/* relay.reshape */
+// TVM_REGISTER_NODE_TYPE(ReshapeAttrs);
```
Comment (icemelon9) [inline src/relay/op/dyn/tensor/transform.cc:37]: remove the comment

File: src/relay/op/tensor/transform.h [hunk 8]
```diff
@@ -38,7 +38,7 @@
 namespace tvm {
 namespace relay {
 
-extern Expr MakeReshape(Expr data, Expr newshape);
+extern Expr MakeReshape(Expr data, Array<Integer> newshape);
```
Comment (icemelon9) [inline src/relay/op/tensor/transform.h:41]:
```suggestion
Expr MakeReshape(Expr data, Array<Integer> newshape);
```

File: python/tvm/relay/op/_tensor_grad.py (hunk 6 above)
Comment (mbrookhart) [reply python/tvm/relay/op/_tensor_grad.py:514]: The numerical approx_grad doesn't seem to work for this case, and it doesn't seem the dynamic version was tested before. I need to look at the autodiff infrastructure more closely, perhaps this should be a second PR?

Comment (icemelon9) [reply python/tvm/relay/op/_tensor_grad.py:514]: Agree.

**Comment by icemelon9** [comment]
Thanks @mbrookhart @lixiaoquan @electriclilies @jroesch

**Pull request merged**